# KPMG Virtual Data Analytics Internship

## Problem scenerio

Here is the background information on your task
Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation, has approached Tony Smith (Partner) in KPMG’s Lighthouse & Innovation Team. Sprocket Central Pty Ltd  is keen to learn more about KPMG’s expertise in its Analytics, Information & Modelling team. 

Smith discusses KPMG’s expertise in this space (you can read more here). In particular, he speaks about how the team can effectively analyse the datasets to help Sprocket Central Pty Ltd grow its business.

Primarily, Sprocket Central Pty Ltd needs help with its customer and transactions data. The organisation has a large dataset relating to its customers, but their team is unsure how to effectively analyse it to help optimise its marketing strategy. 

However, in order to support the analysis, you speak to the Associate Director for some ideas and she advised that “the importance of optimising the quality of customer datasets cannot be underestimated. The better the quality of the dataset, the better chance you will be able to use it drive company growth.”

The client provided KPMG with 3 datasets:

1. Customer Demographic 
2. Customer Addresses
3. Transactions data in the past 3 months

You decide to start the preliminary data exploration and identify ways to improve the quality of Sprocket Central Pty Ltd’s data.


### Task 1:  Data Quality Assessment

In [208]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime

# Read the entire data set
data = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx')

# Seperate the three dataset for further analysis
demographic = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic',skiprows=1)
demographic

/var/folders/wk/q2x1_8p93p98g226j8kx4_nh0000gn/T/ipykernel_4099/1085034433.py:10: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  demographic = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic',skiprows=1)


customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3997         3998          Sarene    Woolley       U   
3998         3999       Patrizius        NaN    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases        DOB               job_title  \
0                                      93 1953-10-12     Executive Secretary   
1                                      81 1980-12-16  Administrative Officer   
2                                      61 1954-01-20      Recruiting Manager   
3                                      33 1961-10-03                     NaN   
4                                      56 1977-05-13           Senior Editor   
...                                   ...        ...                     ...   
3995                                    8 1975-08-09   VP Product Management   
3996                                   87 2001-07-13         Statistician II   
3997                                   60        NaT       Assistant Manager   
3998                                   11 1973-10-24                     NaN   
3999                                   76 1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator  \
0                   Health      Mass Customer                  N   
1       Financial Services      Mass Customer                  N   
2                 Property      Mass Customer                  N   
3                       IT      Mass Customer                  N   
4                      NaN  Affluent Customer                  N   
...                    ...                ...                ...   
3995                Health      Mass Customer                  N   
3996         Manufacturing     High Net Worth                  N   
3997                    IT     High Net Worth                  N   
3998         Manufacturing  Affluent Customer                  N   
3999                   NaN  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                   2018-02-01 00:00:00      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3995                                               -100       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3997                                                NaN       No     NaN  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[4000 rows x 13 columns]

In [209]:
# Similarly import other two sheet data

# Seperate the three dataset for further analysis

cust_address = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerAddress',skiprows=1)
transanction = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='Transactions',skiprows=1)
new_cust_list = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='NewCustomerList',skiprows=1)

/var/folders/wk/q2x1_8p93p98g226j8kx4_nh0000gn/T/ipykernel_4099/3741749760.py:7: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  new_cust_list = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='NewCustomerList',skiprows=1)


### a. Exploring demographic dataset

In [210]:
# Cheking the dataset & shape

demographic

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3997         3998          Sarene    Woolley       U   
3998         3999       Patrizius        NaN    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases        DOB               job_title  \
0                                      93 1953-10-12     Executive Secretary   
1                                      81 1980-12-16  Administrative Officer   
2                                      61 1954-01-20      Recruiting Manager   
3                                      33 1961-10-03                     NaN   
4                                      56 1977-05-13           Senior Editor   
...                                   ...        ...                     ...   
3995                                    8 1975-08-09   VP Product Management   
3996                                   87 2001-07-13         Statistician II   
3997                                   60        NaT       Assistant Manager   
3998                                   11 1973-10-24                     NaN   
3999                                   76 1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator  \
0                   Health      Mass Customer                  N   
1       Financial Services      Mass Customer                  N   
2                 Property      Mass Customer                  N   
3                       IT      Mass Customer                  N   
4                      NaN  Affluent Customer                  N   
...                    ...                ...                ...   
3995                Health      Mass Customer                  N   
3996         Manufacturing     High Net Worth                  N   
3997                    IT     High Net Worth                  N   
3998         Manufacturing  Affluent Customer                  N   
3999                   NaN  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                   2018-02-01 00:00:00      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3995                                               -100       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3997                                                NaN       No     NaN  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[4000 rows x 13 columns]

In [211]:
# Data Profiling (i.e. data types, non null values and columns)

demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [212]:
# Null values in the datset

demographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In the demographic sheet their are 6 column having missing dataset i.e job_industry_category,  job_title,  job_title, DOB,  last_name,  and tenure.

In [213]:
# Checking any duplicate values in the datset

demographic.duplicated().sum() # No duplicate values

0

In [214]:
# Check unique values in each columns;

demographic.nunique()

customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

In [215]:
# Explore columns in Demographic sheet

# 1.Gender

demographic['gender'].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

In [216]:
# Specifying the categories properly:

demographic['gender'] = demographic['gender'].replace('F','Female').replace('M','Male').replace('Femal','Female').replace('U','Unspecified')

demographic['gender'].value_counts()

Female         2039
Male           1873
Unspecified      88
Name: gender, dtype: int64

In [217]:
# 2. past_3_years_bike_related_purchases column

demographic['past_3_years_bike_related_purchases'].value_counts()

16    56
19    56
67    54
20    54
2     50
      ..
8     28
95    27
85    27
86    27
92    24
Name: past_3_years_bike_related_purchases, Length: 100, dtype: int64

In [218]:
# 3. Checking DOB column for any ages above 100

# Get the current date
current_date = datetime.now()       

# Calculate the age by subtracting the DOB from the current date
demographic['age'] = (current_date - demographic['DOB']) / pd.Timedelta(days=365.25)

# Filter rows where age is greater than 100
age_greater_than_100 = demographic[demographic['age'] > 100]

age_greater_than_100

customer_id first_name last_name       gender  \
33           34   Jephthah  Bachmann  Unspecified   

    past_3_years_bike_related_purchases        DOB        job_title  \
33                                   59 1843-12-21  Legal Assistant   

   job_industry_category     wealth_segment deceased_indicator default  \
33                    IT  Affluent Customer                  N     NaN   

   owns_car  tenure         age  
33       No    20.0  179.432637

Here, Jephthah Jephthah has an age of 179 which is practically impossible. Thus, we need to think on imputing it or dropping it.

In [219]:
# 4. Checking job_title 

demographic['job_title'].value_counts()

Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: job_title, Length: 195, dtype: int64

It seems like more of the individula work as a business, tax accountant, social workers, internal auditors & recruiting managers are the prime users.

In [220]:
# 5. Checking job_industry_category 

demographic['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

It seems like more of the individula work for Manufacturing, financial Services, Health are the prime users. Followed by  Retail, Property & IT domain. to use the product. 

In [221]:
# 6. Checking wealth_segment 
 
demographic['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [222]:
# 7. Checking default 

demographic['default'].value_counts()
demographic = demographic.drop('default', axis=1)
demographic

customer_id      first_name  last_name       gender  \
0               1         Laraine  Medendorp       Female   
1               2             Eli    Bockman         Male   
2               3           Arlin     Dearle         Male   
3               4          Talbot        NaN         Male   
4               5  Sheila-kathryn     Calton       Female   
...           ...             ...        ...          ...   
3995         3996         Rosalia   Halgarth       Female   
3996         3997          Blanch     Nisuis       Female   
3997         3998          Sarene    Woolley  Unspecified   
3998         3999       Patrizius        NaN         Male   
3999         4000           Kippy    Oldland         Male   

      past_3_years_bike_related_purchases        DOB               job_title  \
0                                      93 1953-10-12     Executive Secretary   
1                                      81 1980-12-16  Administrative Officer   
2                                      61 1954-01-20      Recruiting Manager   
3                                      33 1961-10-03                     NaN   
4                                      56 1977-05-13           Senior Editor   
...                                   ...        ...                     ...   
3995                                    8 1975-08-09   VP Product Management   
3996                                   87 2001-07-13         Statistician II   
3997                                   60        NaT       Assistant Manager   
3998                                   11 1973-10-24                     NaN   
3999                                   76 1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
0                   Health      Mass Customer                  N      Yes   
1       Financial Services      Mass Customer                  N      Yes   
2                 Property      Mass Customer                  N      Yes   
3                       IT      Mass Customer                  N       No   
4                      NaN  Affluent Customer                  N      Yes   
...                    ...                ...                ...      ...   
3995                Health      Mass Customer                  N       No   
3996         Manufacturing     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   
3998         Manufacturing  Affluent Customer                  N      Yes   
3999                   NaN  Affluent Customer                  N       No   

      tenure        age  
0       11.0  69.625656  
1       16.0  42.447011  
2       15.0  69.351871  
3        7.0  61.650297  
4        8.0  46.041809  
...      ...        ...  
3995    19.0  47.802247  
3996     1.0  21.874800  
3997     NaN        NaN  
3998    10.0  49.592802  
3999    11.0  31.561316  

[4000 rows x 13 columns]

This deafult column is of no importance and can be deleted.

In [223]:
# 8. Checking deceased_indicator 

demographic['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

We have two deceased customer

In [224]:
# 9. Checking owns_car 

demographic['owns_car'].value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

Looks like we have equal number of customers who has a car and who doesnt have a car.

In [225]:
# 10. Checking tenure 

demographic['tenure'].value_counts()

7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
9.0     200
14.0    200
6.0     192
13.0    191
4.0     191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: tenure, dtype: int64

### b. Exploring customer address dataset

In [226]:
# Cheking the dataset & shape

cust_address

customer_id                    address  postcode            state  \
0               1         060 Morning Avenue      2016  New South Wales   
1               2        6 Meadow Vale Court      2153  New South Wales   
2               4         0 Holy Cross Court      4211              QLD   
3               5        17979 Del Mar Point      2448  New South Wales   
4               6           9 Oakridge Court      3216              VIC   
...           ...                        ...       ...              ...   
3994         3999            1482 Hauk Trail      3064              VIC   
3995         4000  57042 Village Green Point      4511              QLD   
3996         4001     87 Crescent Oaks Alley      2756              NSW   
3997         4002           8194 Lien Street      4032              QLD   
3998         4003            320 Acker Drive      2251              NSW   

        country  property_valuation  
0     Australia                  10  
1     Australia                  10  
2     Australia                   9  
3     Australia                   4  
4     Australia                   9  
...         ...                 ...  
3994  Australia                   3  
3995  Australia                   6  
3996  Australia                  10  
3997  Australia                   7  
3998  Australia                   7  

[3999 rows x 6 columns]

In [227]:
# Data Profiling (i.e. data types, non null values and columns)

cust_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [228]:
# Null values in the datset

cust_address.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

No null data

In [229]:
# Checking any duplicate values in the datset

cust_address.duplicated().sum()  # No duplicate values

0

In [230]:
# Check unique values in each columns;

cust_address.nunique()

customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

In [231]:
# Explore columns in Demographic sheet

# 1.State

cust_address['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [232]:
# Specifying the categories properly:

cust_address['state']= cust_address['state'].replace('Victoria','VIC').replace('New South Wales','NSW')

cust_address['state'].value_counts()

NSW    2140
VIC    1021
QLD     838
Name: state, dtype: int64

We have maximum users from 'New South Wales' followed by 'Victoria' & 'Queensland'.

In [233]:
# 2. checking property_valuation

cust_address['property_valuation'].value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

### c. Exploring Transaction dataset

In [234]:
# Cheking the dataset & shape

transanction

transaction_id  product_id  customer_id transaction_date  online_order  \
0                   1           2         2950       2017-02-25           0.0   
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost  product_first_sold_date  
0           71.49          53.62                  41245.0  
1         2091.47         388.92                  41701.0  
2         1793.43         248.82                  36361.0  
3         1198.46         381.10                  36145.0  
4         1765.30         709.48                  42226.0  
...           ...            ...                      ...  
19995     2005.66        1203.40                  37823.0  
19996      416.98         312.74                  35560.0  
19997     1636.90          44.71                  40410.0  
19998      227.88         136.73                  38216.0  
19999     1775.81        1580.47                  36334.0  

[20000 rows x 13 columns]

In [235]:
# Data Profiling (i.e. data types, non null values and columns)

transanction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [236]:
# Checking any duplicate values in the datset

transanction.duplicated().sum() # No duplicate values

0

In [237]:
# Null values in the datset

transanction.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In the transanction sheet their are 7 column having missing dataset i.e online_order,  brand,  product_line, product_class,  product_size, standard_cost and product_first_sold_date.

In [238]:
# Converting product_first_sold_date in date format

transanction['product_first_sold_date'] = pd.to_datetime(transanction['product_first_sold_date'])
transanction

transaction_id  product_id  customer_id transaction_date  online_order  \
0                   1           2         2950       2017-02-25           0.0   
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost       product_first_sold_date  
0           71.49          53.62 1970-01-01 00:00:00.000041245  
1         2091.47         388.92 1970-01-01 00:00:00.000041701  
2         1793.43         248.82 1970-01-01 00:00:00.000036361  
3         1198.46         381.10 1970-01-01 00:00:00.000036145  
4         1765.30         709.48 1970-01-01 00:00:00.000042226  
...           ...            ...                           ...  
19995     2005.66        1203.40 1970-01-01 00:00:00.000037823  
19996      416.98         312.74 1970-01-01 00:00:00.000035560  
19997     1636.90          44.71 1970-01-01 00:00:00.000040410  
19998      227.88         136.73 1970-01-01 00:00:00.000038216  
19999     1775.81        1580.47 1970-01-01 00:00:00.000036334  

[20000 rows x 13 columns]

In [239]:
# Check unique values in each columns;

transanction.nunique()

transaction_id             20000
product_id                   101
customer_id                 3494
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   296
standard_cost                103
product_first_sold_date      100
dtype: int64

In [240]:
# Explore columns in Demographic sheet

# 1. online_order

transanction['online_order'].value_counts()

1.0    9829
0.0    9811
Name: online_order, dtype: int64

In [241]:
# 2. order_status

transanction['order_status'].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [242]:
# 3. brand

transanction['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [243]:
# 4. product_line

transanction['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [244]:
# 5. product_class

transanction['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [245]:
# 6. product_size

transanction['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

### d. New Customer List dataset

In [246]:
# Cheking the dataset & shape

new_cust_list

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0      Chickie    Brister    Male                                   86   
1        Morly     Genery    Male                                   69   
2      Ardelis  Forrester  Female                                   10   
3       Lucine      Stutt  Female                                   64   
4      Melinda     Hadlee  Female                                   34   
..         ...        ...     ...                                  ...   
995  Ferdinand  Romanetti    Male                                   60   
996       Burk    Wortley    Male                                   22   
997   Melloney      Temby  Female                                   17   
998     Dickie    Cubbini    Male                                   30   
999      Sylas    Duffill    Male                                   56   

           DOB                     job_title job_industry_category  \
0   1957-07-12               General Manager         Manufacturing   
1   1970-03-22           Structural Engineer              Property   
2   1974-08-28        Senior Cost Accountant    Financial Services   
3   1979-01-28    Account Representative III         Manufacturing   
4   1965-09-21             Financial Analyst    Financial Services   
..         ...                           ...                   ...   
995 1959-10-07                     Paralegal    Financial Services   
996 2001-10-17        Senior Sales Associate                Health   
997 1954-10-05  Budget/Accounting Analyst IV    Financial Services   
998 1952-12-17             Financial Advisor    Financial Services   
999 1955-10-02           Staff Accountant IV              Property   

        wealth_segment deceased_indicator owns_car  ...  state    country  \
0        Mass Customer                  N      Yes  ...    QLD  Australia   
1        Mass Customer                  N       No  ...    NSW  Australia   
2    Affluent Customer                  N       No  ...    VIC  Australia   
3    Affluent Customer                  N      Yes  ...    QLD  Australia   
4    Affluent Customer                  N       No  ...    NSW  Australia   
..                 ...                ...      ...  ...    ...        ...   
995  Affluent Customer                  N       No  ...    NSW  Australia   
996      Mass Customer                  N       No  ...    NSW  Australia   
997  Affluent Customer                  N      Yes  ...    QLD  Australia   
998      Mass Customer                  N      Yes  ...    QLD  Australia   
999      Mass Customer                  N      Yes  ...    NSW  Australia   

     property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                     6        0.56      0.7000     0.875000     0.743750   
1                    11        0.89      0.8900     1.112500     0.945625   
2                     5        1.01      1.0100     1.010000     1.010000   
3                     1        0.87      1.0875     1.087500     1.087500   
4                     9        0.52      0.5200     0.650000     0.650000   
..                  ...         ...         ...          ...          ...   
995                   7        0.79      0.7900     0.790000     0.790000   
996                  10        0.76      0.7600     0.950000     0.807500   
997                   2        0.85      1.0625     1.062500     1.062500   
998                   2        1.09      1.3625     1.362500     1.158125   
999                   9        0.47      0.5875     0.734375     0.624219   

     Unnamed: 20  Rank     Value  
0              1     1  1.718750  
1              1     1  1.718750  
2              1     1  1.718750  
3              4     4  1.703125  
4              4     4  1.703125  
..           ...   ...       ...  
995          996   996  0.374000  
996          997   997  0.357000  
997          997   997  0.357000  
998          997   997  0.357000  
999         1000  1000  0.340000  

[10

We have some unnamed columns, which need some clarification if thats important or else we can go ahead and drop it.

In [247]:
# Data Profiling (i.e. data types, non null values and columns)

new_cust_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [248]:
# Null values in the datset

new_cust_list.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Unnamed: 16                              0
Unnamed: 17                              0
Unnamed: 18                              0
Unnamed: 19                              0
Unnamed: 20                              0
Rank                                     0
Value                                    0
dtype: int6

There are 4 column which have null values i.e. last_name, DOB, job_title, and job_industry_category.

In [249]:
# Checking any duplicate values in the datset

new_cust_list.duplicated().sum() # No duplicate values

0

In [250]:
# Check unique values in each columns;

new_cust_list.nunique()

first_name                              940
last_name                               961
gender                                    3
past_3_years_bike_related_purchases     100
DOB                                     958
job_title                               184
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        1
owns_car                                  2
tenure                                   23
address                                1000
postcode                                522
state                                     3
country                                   1
property_valuation                       12
Unnamed: 16                              71
Unnamed: 17                             132
Unnamed: 18                             183
Unnamed: 19                             321
Unnamed: 20                             324
Rank                                    324
Value                           

In [251]:
# Explore columns in Demographic sheet

# 1.Gender
new_cust_list['gender'].value_counts()

# Specifying the categories properly:
new_cust_list['gender'] = demographic['gender'].replace('U','Unspecified')

new_cust_list['gender'].value_counts()

Female         511
Male           465
Unspecified     24
Name: gender, dtype: int64

In [252]:
# 2. Checking DOB column for any ages above 100

# Get the current date
current_date = datetime.now()       

# Calculate the age by subtracting the DOB from the current date
new_cust_list['age'] = (current_date - new_cust_list['DOB']) / pd.Timedelta(days=365.25)

# Filter rows where age is greater than 100
age_greater_than_100 = new_cust_list[new_cust_list['age'] > 100]

age_greater_than_100

Empty DataFrame
Columns: [first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, address, postcode, state, country, property_valuation, Unnamed: 16, Unnamed: 17, Unnamed: 18, Unnamed: 19, Unnamed: 20, Rank, Value, age]
Index: []

[0 rows x 24 columns]

None of the customer age are greater than 100

In [253]:
# 3. Checking job_title 

new_cust_list['job_title'].value_counts()

Associate Professor         15
Environmental Tech          14
Software Consultant         14
Chief Design Engineer       13
Assistant Manager           12
                            ..
Accountant II                1
Programmer IV                1
Administrative Officer       1
Accounting Assistant III     1
Web Developer I              1
Name: job_title, Length: 184, dtype: int64

It seems like more of the individula work as a Associate Professor, Environmental Tech, Software Consultant, Chief Design Engineer & Assistant Manager  are the prime users.

In [254]:
# 4. Checking job_industry_category 

new_cust_list['job_industry_category'].value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

It seems like more of the individula work for Manufacturing, Financial Services, Health are the prime users. Followed by  Retail, Property & IT domain. to use the product. 

In [255]:
# 5. Checking wealth_segment 
 
new_cust_list['wealth_segment'].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [256]:
# 6. Checking deceased_indicator 

new_cust_list['deceased_indicator'].value_counts()

N    1000
Name: deceased_indicator, dtype: int64

We have no customer status as deceased

In [257]:
# 7. Checking owns_car 

new_cust_list['owns_car'].value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [258]:
# 8. Checking tenure 

demographic['tenure'].value_counts()

7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
9.0     200
14.0    200
6.0     192
13.0    191
4.0     191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: tenure, dtype: int64

In [259]:
# 8. Checking State 

new_cust_list['state'].value_counts()

NSW    506
VIC    266
QLD    228
Name: state, dtype: int64

In [260]:
# 8. Checking property_valuation 

new_cust_list['property_valuation'].value_counts()

9     173
8     162
7     138
10    116
6      70
11     62
5      57
4      53
3      51
12     46
2      42
1      30
Name: property_valuation, dtype: int64

### Task 2:  Data Insights